editor/plugins/visual_shape_2d_editor_plugin.cpp

```diff
@@ -0,0 +1,292 @@
+		EditorToaster::get_singleton()->popup_str(TTR("Invalid geometry, can't replace by mesh."), EditorToaster::SEVERITY_ERROR);
+		return;
+	}
+
+	Vector<int> poly = Geometry2D::triangulate_polygon(points);
+
+	Ref<ArrayMesh> mesh;
+	mesh.instantiate();
+
+	Array a;
+	a.resize(Mesh::ARRAY_MAX);
+	a[Mesh::ARRAY_VERTEX] = points;
+	a[Mesh::ARRAY_TEX_UV] = visual_shape_2d->get_uvs();
+	a[Mesh::ARRAY_INDEX] = poly;
+
+	mesh->add_surface_from_arrays(Mesh::PRIMITIVE_TRIANGLES, a, Array(), Dictionary(), Mesh::ARRAY_FLAG_USE_2D_VERTICES);
+
+	MeshInstance2D *mesh_instance = memnew(MeshInstance2D);
+	mesh_instance->set_mesh(mesh);
+
+	EditorUndoRedoManager *ur = EditorUndoRedoManager::get_singleton();
+	ur->create_action(TTR("Convert to MeshInstance2D"), UndoRedo::MERGE_DISABLE, visual_shape_2d);
+	SceneTreeDock::get_singleton()->replace_node(visual_shape_2d, mesh_instance);
+	ur->commit_action(false);
+}
+
+void VisualShape2DEditor::_convert_to_polygon_2d_node() {
+	PackedVector2Array points = visual_shape_2d->get_points();
+	if (points.is_empty()) {
+		EditorToaster::get_singleton()->popup_str(TTR("Invalid geometry, can't create polygon."), EditorToaster::SEVERITY_ERROR);
+		return;
+	}
+
+	int total_point_count = points.size();
+	Point2 offset = visual_shape_2d->get_offset();
+
+	Polygon2D *polygon_2d_instance = memnew(Polygon2D);
+
+	polygon_2d_instance->set_color(visual_shape_2d->get_color());
+	polygon_2d_instance->set_offset(offset);
+	polygon_2d_instance->set_antialiased(visual_shape_2d->is_antialiased());
+
+	PackedVector2Array vertices;
+	vertices.resize(total_point_count);
+	Vector2 *vertices_write = vertices.ptrw();
+
+	PackedInt32Array index_array;
+	index_array.resize(total_point_count);
+	int *index_write = index_array.ptrw();
+
+	for (int i = 0; i < total_point_count; i++) {
+		vertices_write[i] = points[i] - offset;
+		index_write[i] = i;
+	}
+
+	Array polys;
+	polys.push_back(index_array);
+
+	PackedVector2Array uvs = Transform2D(0, visual_shape_2d->get_size(), 0, Point2(0, 0)).xform(visual_shape_2d->get_uvs());
+
+	polygon_2d_instance->set_polygon(vertices);
+	polygon_2d_instance->set_uv(uvs);
+	polygon_2d_instance->set_polygons(polys);
+
+	EditorUndoRedoManager *ur = EditorUndoRedoManager::get_singleton();
+	ur->create_action(TTR("Convert to Polygon2D"), UndoRedo::MERGE_DISABLE, visual_shape_2d);
+	SceneTreeDock::get_singleton()->replace_node(visual_shape_2d, polygon_2d_instance);
+	ur->commit_action(false);
+}
+
+void VisualShape2DEditor::_create_collision_shape_2d_node() {
+	CollisionShape2D *collision_shape_2d_instance = memnew(CollisionShape2D);
+	Size2 size = visual_shape_2d->get_size();
+
+	switch (visual_shape_2d->get_shape_type()) {
+		case VisualShape2D::SHAPE_RECTANGLE: {
+			RectangleShape2D *shape = memnew(RectangleShape2D);
+			shape->set_size(size);
+			collision_shape_2d_instance->set_shape(shape);
+			collision_shape_2d_instance->translate(visual_shape_2d->get_offset());
+		} break;
+		case VisualShape2D::SHAPE_CIRCLE: {
```

Comment:
This might need a special case for ovals. It's the only shape that doesn't get accurate collider.

Reply:
I could calculate the eccentricity and make a ConvexPolygonShape2D in some cases, but I didn't want to arbitrarily choose some threshold. I also want to have it use CircleShape2D where possible, since it is better for physics.
I can play around with thresholds and see.

Reply:
Eccentricity wasn't very helpful, for now I went with a 10% difference in size.
If it still feels off, I could instead do what MeshInstance3DEditor does and have a menu and let the user choose whether to make it a shape or a convex polygon. This would also allow for small resolution shapes (like hexagon) to have a polygon option.